### DISCUSSION TITLE
[Lesson 6] A better UX about wallet management

### DISCUSSION THREAD

**Discussion by Osneue** [Cyfrin/foundry-full-course-cu#945] [Ideas]
I find [an issue about UX of keystore](https://github.com/foundry-rs/foundry/issues/1869#issue-1263794111) through Partick's solidity course. 

And I keep an eye out for this new feature and finally it was merged at [here](https://github.com/foundry-rs/foundry/pull/5551#issue-1838407437).

I try to test this new feature.

I import my private key (from Ganache local blockchain) by typing
```
cast wallet import gan-wall --interactive
```

and I get the output
```
Enter private key:
Enter password:
`gan-wall` keystore was saved successfully. Address: 0xdcc0ed94411272c1eb99067f34dfb0aefc511709
```

The address `0xdcc0ed94411272c1eb99067f34dfb0aefc511709` listed above correspondes to the private key I input. 

So I think the saved keystore has recorded the address corresponding to the private key.

It all works well up to now.

And then I try to deploy my smart contract using script by typing
```
forge script script/DeploySimpleStorage.s.sol --rpc-url $GANACHE_RPC_URL --broadcast --account gan-wall
```
But I get an error
```
[⠒] Compiling...
No files changed, compilation skipped
Script ran successfully.

== Return ==
0: contract SimpleStorage 0x34A1D3fff3958843C43aD80F30b94c510645C316

EIP-3855 is not supported in one or more of the RPCs used.
Unsupported Chain IDs: 1337.
Contracts deployed with a Solidity version equal or higher than 0.8.20 might not work properly.
For more information, please see https://eips.ethereum.org/EIPS/eip-3855

## Setting up (1) EVMs.

==========================

Chain 1337

Estimated gas price: 4.380813908 gwei

Estimated total gas used for script: 464097

Estimated amount required: 0.002033122592261076 ETH

==========================

###
Finding wallets for all the necessary addresses...
Enter keystore password:

Transactions saved to: /home/sjz/solidity-course/patrick/foundry-simple-storage-f23/broadcast/DeploySimpleStorage.s.sol/1337/run-latest.json

Sensitive values saved to: /home/sjz/solidity-course/patrick/foundry-simple-storage-f23/cache/DeploySimpleStorage.s.sol/1337/run-latest.json

Error:

You seem to be using Foundry's default sender. Be sure to set your own --sender.
No associated wallet for addresses: {0x1804c8ab1f12e6bbf3894d4083f33e07309d1f38}. Unlocked wallets: [0xdcc0ed94411272c1eb99067f34dfb0aefc511709]
```

Obviously it is caused by the wrong address, so I retry it by typing
```
forge script script/DeploySimpleStorage.s.sol --rpc-url $GANACHE_RPC_URL --broadcast --account gan-wall --sender 0xdCC0eD94411272c1eB99067F34dFb0AEfc511709
```
Finally, everything goes well.

And what puzzles me is

- **Since I have already specified --account, why do I need to specify --sender?**

- **Why does the --account option not automatically select the corresponding address?**

I am brand new to this field, please forgive me if my question is stupid.

And please correct me if there is anything wrong with my formatting. @PatrickAlphaC or anybody else

**Comment by 0xusmanf** [comment]
I think specifying `--account` should automatically select the corresponding address as the sender because technically `--account` is the replacement for `--private-key`, and when you use `--private-key` you don't have to explicitly specify `--sender` as it automatically defaults to it, same should be the case with `--account`.

Would love to hear what @PatrickAlphaC thinks about it.

**Reply by Osneue** [reply]
Yes, I totally agree with you.

**Reply by PatrickAlphaC** [reply]
I agree. Can we make an issue in foundry with this? Perhaps we can reach out to https://twitter.com/rplusq 

If they are game, I'm happy to sponsor this work .

**Reply by 0xusmanf** [reply]
@rplusq Please check this thread.

**Reply by Osneue** [reply]
I’ve made an issue https://github.com/foundry-rs/foundry/issues/6034#issue-1940954508

**Reply by rplusq** [reply]
Thanks for raising this, will work on it.

**Reply by rplusq** [reply]
Tried to reproduce, didn't get it as easily. Did you set any specific env vars before? @Osneue

**Reply by 0xusmanf** [reply]
I reproduced it on this repo https://github.com/Cyfrin/foundry-simple-storage-f23/tree/main, The only env var I had to set was `$GANACHE_RPC_URL` in my `.env`.

**Reply by Osneue** [reply]
@rplusq 
## what I do
I restart a new termial without any env var set
I input the command 
```
forge script script/DeployFundMe.s.sol --rpc-url https://xxxxxx --broadcast --account xxxx
```
and get these
```
###
Finding wallets for all the necessary addresses...
Enter keystore password:

Transactions saved to: /home/sjz/solidity-course/sjz/foundry-fund-me-f23/broadcast/DeployFundMe.s.sol/11155111/run-latest.json

Sensitive values saved to: /home/sjz/solidity-course/sjz/foundry-fund-me-f23/cache/DeployFundMe.s.sol/11155111/run-latest.json

Error:

You seem to be using Foundry's default sender. Be sure to set your own --sender.
No associated wallet for addresses: {0x1804c8ab1f12e6bbf3894d4083f33e07309d1f38}. Unlocked wallets: [0xe5ded0a256eab71b2298bae64b0c02a892f1bced]
```
It is worth noting that `Unlocked wallets: [0xe5ded0a256eab71b2298bae64b0c02a892f1bced]`, that's the address corresponding to my private key stored in wallet, why is it makred as `Unlocked wallet` ?
## additional context
ubuntu 20.04 in wsl2

**Reply by Osneue** [reply]
@PatrickAlphaC Please provide further details regarding the sponsorship.

**Reply by rplusq** [reply]
Got to reproduce now, working on it!

**Reply by Stachelabs** [reply]
Hi everyone, I don't know if it has been improved since, but I use your solution --account <NAME OF THE ACCOUNT> --sender <PUBLIC ADDRESS CORRESPONDING TO THE NAME OF THE ACCOUNT> and it works.. hopefully we will be able to have a shortcut soon with just --account but it is working
